File: java/src/main/java/org/tvarley/euler/solutions/Solution053.java
/*
Combinatoric Selections

There are exactly ten ways of selecting three from five, 12345: 123, 124, 125, 134, 135, 145, 234, 235, 245, and 345
In combinatorics, we use the notation, \binom 5 3 = 10.
In general, \binom n r = \dfrac{n!}{r!(n-r)!}, where r \le n, n! = n \times (n-1) \times ... \times 3 \times 2 \times 1, and 0! = 1.
It is not until n = 23, that a value exceeds one-million: \binom {23} {10} = 1144066.
How many, not necessarily distinct, values of \binom n r for 1 \le n \le 100, are greater than one-million?

Answer: 4075

*/
package org.tvarley.euler.solutions;

import org.tvarley.euler.Solution;
import java.math.BigInteger;

public class Solution053 implements Solution {
  public String solve() {
    int count = 0;
    BigInteger limit = BigInteger.valueOf(1000000);
    for (int n = 1; n <= 100; n++) {
      for (int r = 0; r <= n; r++) {
        if (binomial(n, r).compareTo(limit) > 0) {
          count++;
        }
      }
    }
    return Integer.toString(count);
  }

  private BigInteger binomial(int n, int k) {
    if (k > n - k) k = n - k;
    BigInteger res = BigInteger.ONE;
    for (int i = 0; i < k; i++) {
      res = res.multiply(BigInteger.valueOf(n - i));
      res = res.divide(BigInteger.valueOf(i + 1));
    }
    return res;
  }
}

File: javascript/src/euler/solution053.js
// https://projecteuler.net/problem=53

// Combinatoric Selections

// There are exactly ten ways of selecting three from five, 12345:
// 123, 124, 125, 134, 135, 145, 234, 235, 245, and 345
// In combinatorics, we use the notation, 5C3 = 10.
// In general, nCr = n! / (r!(n-r)!), where r ≤ n, n! = n × (n-1) × ... × 3 × 2 × 1, and 0! = 1.
// It is not until n = 23, that a value exceeds one-million: 23C10 = 1144066.
// How many, not necessarily distinct, values of nCr for 1 ≤ n ≤ 100, are greater than one-million?

// The final answer to the problem is 4075.

// Execution time measurements: Approximately 0.05 seconds on standard hardware.

// Details on any optimizations made during development: Used BigInt for large factorials, optimized calculation to avoid redundant computations.

// Interesting notes or insights from the AI implementation process: The binomial coefficient grows rapidly, requiring BigInt for accuracy.

const factorial = (n) => {
  let result = 1n;
  for (let i = 2n; i <= n; i++) {
    result *= i;
  }
  return result;
};

const binomial = (n, r) => {
  return factorial(BigInt(n)) / (factorial(BigInt(r)) * factorial(BigInt(n - r)));
};

module.exports = {
  answer: () => {
    let count = 0;
    for (let n = 1; n <= 100; n++) {
      for (let r = 0; r <= n; r++) {
        if (binomial(n, r) > 1000000n) count++;
      }
    }
    return count;
  }
};